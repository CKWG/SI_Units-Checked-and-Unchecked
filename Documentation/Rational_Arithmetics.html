<?xml version="1.0" encoding="ISO-8859-1" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
                      "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">

<head>
<link rel="shortcut icon" type="image/x-icon" href="Ada.ico" />
<title>Rational_Arithmetics</title>
<!-- Version  Date
      1.0  10.01.2006
      1.1  01.03.2006  pragma Pure: Put Image and Value in child
      1.2  10.08.2018  Ada.ico
      1.3  21.08.2018  Validated
//-->
</head>

<body>

<h2 align="center">Rational_Arithmetics</h2>

<p>The package Rational_Arithmetics provides operations for rational (n/m) and
mixed (i + n/m) numbers, where i, n and m are whole numbers.
All operations resulting in rational numbers are implemented such that the
result is reduced and has a positive denominator.</p>

<h3 align="left">Specification</h3>

<pre><b>package</b> Rational_Arithmetics <b>is

  pragma</b> Pure;
  <b>pragma</b> Elaborate_Body;

  -- <i>Whole numbers</i>

  <b>type</b> Whole <b>is new</b> Integer;
  <b>subtype</b> Natural_Whole  <b>is</b> Whole <b>range</b> 0 .. Whole'Last;
  <b>subtype</b> Positive_Whole <b>is</b> Whole <b>range</b> 1 .. Whole'Last;

  -- <i>Undefine unwanted operations</i>

  <b>function</b> &quot;+&quot; (Right: Whole) <b>return</b> Whole <b>is abstract</b>;
  <b>function</b> &quot;-&quot; (Right: Whole) <b>return</b> Whole <b>is abstract</b>;

  <b>function</b> &quot;/&quot; (Left, Right: Whole) <b>return</b> Whole <b>is abstract</b>;

  -- <i>Rational numbers</i>

  <b>type</b> Rational <b>is private</b>;

  <b>function</b> Numerator   (R: Rational) <b>return</b> Whole;
  <b>function</b> Denominator (R: Rational) <b>return</b> Positive_Whole;

  -- <i>Constructors</i>

  <b>function</b> &quot;+&quot; (Right: Whole) <b>return</b> Rational;
  <b>function</b> &quot;-&quot; (Right: Whole) <b>return</b> Rational;

  <b>function</b> &quot;/&quot; (Left, Right: Whole) <b>return</b> Rational;

  -- <i>Rational operations</i>

  <b>function</b> &quot;abs&quot; (Right: Rational) <b>return</b> Rational;</pre>

<pre>  <b>function</b> &quot;+&quot; (Right: Rational) <b>return</b> Rational;
  <b>function</b> &quot;-&quot; (Right: Rational) <b>return</b> Rational;

  <b>function</b> &quot;+&quot; (Left, Right: Rational) <b>return</b> Rational;
  <b>function</b> &quot;-&quot; (Left, Right: Rational) <b>return</b> Rational;

  <b>function</b> &quot;*&quot; (Left, Right: Rational) <b>return</b> Rational;
  <b>function</b> &quot;/&quot; (Left, Right: Rational) <b>return</b> Rational;

  -- <i>Relational</i>

  <b>function</b> &quot;=&quot; (Left: Whole   ; Right: Rational) <b>return</b> Boolean;
  <b>function</b> &quot;=&quot; (Left: Rational; Right: Whole   ) <b>return</b> Boolean;

  <b>function</b> &quot;&lt;&quot;  (Left, Right: Rational) <b>return</b> Boolean;
  <b>function</b> &quot;&lt;=&quot; (Left, Right: Rational) <b>return</b> Boolean;
  <b>function</b> &quot;&gt;=&quot; (Left, Right: Rational) <b>return</b> Boolean;
  <b>function</b> &quot;&gt;&quot;  (Left, Right: Rational) <b>return</b> Boolean;

  <b>function</b> &quot;&lt;&quot;  (Left: Whole   ; Right: Rational) <b>return</b> Boolean;
  <b>function</b> &quot;&lt;&quot;  (Left: Rational; Right: Whole   ) <b>return</b> Boolean;
  <b>function</b> &quot;&lt;=&quot; (Left: Whole   ; Right: Rational) <b>return</b> Boolean;
  <b>function</b> &quot;&lt;=&quot; (Left: Rational; Right: Whole   ) <b>return</b> Boolean;
  <b>function</b> &quot;&gt;=&quot; (Left: Whole   ; Right: Rational) <b>return</b> Boolean;
  <b>function</b> &quot;&gt;=&quot; (Left: Rational; Right: Whole   ) <b>return</b> Boolean;
  <b>function</b> &quot;&gt;&quot;  (Left: Whole   ; Right: Rational) <b>return</b> Boolean;
  <b>function</b> &quot;&gt;&quot;  (Left: Rational; Right: Whole   ) <b>return</b> Boolean;

  <b>function</b> Max (Left, Right: Rational) <b>return</b> Rational;
  <b>function</b> Min (Left, Right: Rational) <b>return</b> Rational;

  <b>function</b> Max (Left: Whole   ; Right: Rational) <b>return</b> Rational;
  <b>function</b> Max (Left: Rational; Right: Whole   ) <b>return</b> Rational;
  <b>function</b> Min (Left: Whole   ; Right: Rational) <b>return</b> Rational;
  <b>function</b> Min (Left: Rational; Right: Whole   ) <b>return</b> Rational;

  -- <i>Mixed operations</i>

  <b>function</b> &quot;+&quot; (Left: Whole   ; Right: Rational) <b>return</b> Rational;
  <b>function</b> &quot;+&quot; (Left: Rational; Right: Whole   ) <b>return</b> Rational;
  <b>function</b> &quot;-&quot; (Left: Whole   ; Right: Rational) <b>return</b> Rational;
  <b>function</b> &quot;-&quot; (Left: Rational; Right: Whole   ) <b>return</b> Rational;

  <b>function</b> &quot;*&quot; (Left: Whole   ; Right: Rational) <b>return</b> Rational;
  <b>function</b> &quot;*&quot; (Left: Rational; Right: Whole   ) <b>return</b> Rational;
  <b>function</b> &quot;/&quot; (Left: Whole   ; Right: Rational) <b>return</b> Rational;
  <b>function</b> &quot;/&quot; (Left: Rational; Right: Whole   ) <b>return</b> Rational;

<b>private</b>

  <i>Not shown</i>

<b>end</b> Rational_Arithmetics;</pre>

<p>The operations</p>

<pre>  <b>function</b> &quot;+&quot; (Right: Whole) <b>return</b> Whole <b>is abstract</b>;
  <b>function</b> &quot;-&quot; (Right: Whole) <b>return</b> Whole <b>is abstract</b>;

  <b>function</b> &quot;/&quot; (Left, Right: Whole) <b>return</b> Whole <b>is abstract</b>;</pre>

<p>are undefined because they are needed as constructors:</p>

<pre>  <b>function</b> &quot;+&quot; (Right: Whole) <b>return</b> Rational;
  <b>function</b> &quot;-&quot; (Right: Whole) <b>return</b> Rational;

  <b>function</b> &quot;/&quot; (Left, Right: Whole) <b>return</b> Rational;</pre>

<p>All binary operators are defined such that they can be applied on rational
numbers alone or on a rational and a whole number in any sequence.</p>

<p><a href="Table_of_Contents.html">Table of Contents</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<a href="Rational_Arithmetics-Strings.html">Rational_Arithmetics.Strings</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<a href="Rational_Arithmetics-Text_IO.html">Rational_Arithmetics.Text_IO</a></p>

<p align="center"><a href="http://validator.w3.org/check?uri=referer">
   <img src="valid-xhtml10.png" alt="Valid XHTML 1.0 Transitional" height="31" width="88" /></a></p>

</body>

</html>
