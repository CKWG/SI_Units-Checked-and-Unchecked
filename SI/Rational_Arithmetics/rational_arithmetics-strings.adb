------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2006, 2018 Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Author's homepage and email address:
--   http://www.christ-usch-grein.homepage.t-online.de/
--   Christ-Usch.Grein@T-Online.de
------------------------------------------------------------------------------

with Ada.Strings.Fixed;

package body Rational_Arithmetics.Strings is

  --====================================================================
  -- Author    Christoph Grein
  -- Version   1.0
  -- Date      27 February 2006
  --====================================================================
  --
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    1.0  27.02.2006
  --====================================================================

  function Image (X: Rational) return String is
    N: constant String := Whole'Image (X.Numerator);
    D: constant String := Whole'Image (X.Denominator);
  begin
    if X.Denominator = 1 then
      return N;
    else
      return N & '/' & D (2 .. D'Last);
    end if;
  end Image;

  function Value (X: String) return Rational is
    use Ada.Strings;
    Y: String renames Fixed.Trim (X, Side => Both);
    -- Append '/' so that it is always found:
    S: Natural := Fixed.Index (Y & '/', "/");
    N: constant Whole:= Whole'Value (Y (Y'First .. S-1));
    D: Positive_Whole;
  begin
    if Y (S-1) = ' ' then  -- blank before /
      raise Constraint_Error;
    end if;
    if S > Y'Last then
      return (N, 1);
    elsif Y (S+1) = ' ' then  -- blank after /
      raise Constraint_Error;
    end if;
    D := Whole'Value (Y (S+1 .. Y'Last));
    return N/D;
  end Value;

end Rational_Arithmetics.Strings;
