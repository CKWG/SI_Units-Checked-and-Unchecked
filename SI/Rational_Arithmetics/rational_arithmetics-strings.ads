------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2006, 2018 Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Author's homepage and email address:
--   http://www.christ-usch-grein.homepage.t-online.de/
--   Christ-Usch.Grein@T-Online.de
------------------------------------------------------------------------------

package Rational_Arithmetics.Strings is

  --====================================================================
  -- Author    Christoph Grein
  -- Version   1.1
  -- Date      29 July 2018
  --====================================================================
  -- Image returns the argument as a string.
  -- If the denominator is 1, the string has the format "sn", else it
  -- has the format "sn/m" where there is no white space around '/',
  -- and the sign 's' is blank for non-negative numbers.
  -- Value returns the argument into a rational number ignoring any
  -- leading or trailing blanks.
  -- It raises Constraint_Error if the string does not have the
  -- aforementioned format.
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    1.0  27.02.2006
  --  C.G.    1.0  29.07.2018 Preelaborate
  --====================================================================

  pragma Preelaborate;
  pragma Elaborate_Body;

  function Image (X: Rational) return String;
  function Value (X: String  ) return Rational;

end Rational_Arithmetics.Strings;
