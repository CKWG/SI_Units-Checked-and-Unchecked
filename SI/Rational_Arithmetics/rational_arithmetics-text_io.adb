------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2002, 2018 Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Author's homepage and email address:
--   http://home.T-Online.de/home/Christ-Usch.Grein/
--   Christ-Usch.Grein@T-Online.de
------------------------------------------------------------------------------

with Ada.Integer_Text_IO;
use  Ada.Integer_Text_IO;

package body Rational_Arithmetics.Text_IO is

  --====================================================================
  -- Author    Christoph Grein
  -- Version   1.1
  -- Date      30 June 2002
  --====================================================================
  --
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    0.0  27.06.2002 Only Put operations
  --  C.G.    1.0  30.06.2002 Done
  --  C.G.    1.1  18.07.2002 Interprete number read as integer when
  --                          following character is not '/'
  --====================================================================

  procedure Get (Item: out Rational) is
  begin
    Get (Current_Input, Item);
  end Get;

  procedure Put (Item: in Rational) is
  begin
    Put (Current_Output, Item);
  end Put;

  procedure Get (File: in File_Type; Item: out Rational) is
    N, D: Whole;
    C   : Character;
    EoL : Boolean;
  begin
    Get (File, Integer (N));
    Look_Ahead (File, C, EoL);
    if EoL or else C /= '/' then
      Item := +N;
      return;
    end if;
    Get (File, C);  -- consume '/'
    Look_Ahead (File, C, EoL);
    if EoL or else C not in '0' .. '9' then
      raise Data_Error;
    end if;
    Get (File, Integer (D));
    Item := N/D;
  end Get;

  procedure Put (File: in File_Type; Item: in Rational) is
  begin
    if Item.Denominator = 1 then
      Put (File, Whole'Image (Item.Numerator));
    else
      declare
        D: constant String := Whole'Image (Item.Denominator);
      begin
        Put (File, Whole'Image (Item.Numerator) & '/' & D (2..D'Last));
      end;
    end if;
  end Put;

end Rational_Arithmetics.Text_IO;
