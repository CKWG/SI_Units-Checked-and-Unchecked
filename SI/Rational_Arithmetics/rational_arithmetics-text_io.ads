------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2002, 2018 Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Author's homepage and email address:
--   http://home.T-Online.de/home/Christ-Usch.Grein/
--   Christ-Usch.Grein@T-Online.de
------------------------------------------------------------------------------

with Ada.Text_IO;
use  Ada.Text_IO;

package Rational_Arithmetics.Text_IO is

  --====================================================================
  -- Author    Christoph Grein
  -- Version   1.0
  -- Date      5 July 2002
  --====================================================================
  -- Put and Get work like the conventional operations in Text_IO for
  -- numbers.
  -- If the denominator is 1, put writes the number in the format "sn",
  -- else writes it in the format "sn/m" where there is no white space
  -- around '/', and the sign 's' is blank for non-negative numbers.
  -- Get reads the number and raises Data_Error if it does not have the
  -- aforementioned format.
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    1.0  05.07.2002
  --====================================================================

  procedure Get (Item: out Rational);
  procedure Put (Item: in  Rational);

  procedure Get (File: in File_Type; Item: out Rational);
  procedure Put (File: in File_Type; Item: in  Rational);

end Rational_Arithmetics.Text_IO;
