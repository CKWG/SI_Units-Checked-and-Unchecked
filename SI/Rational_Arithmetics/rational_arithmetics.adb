------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2002, 2005, 2006, 2018 Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Author's homepage and email address:
--   http://www.christ-usch-grein.homepage.t-online.de/
--   Christ-Usch.Grein@T-Online.de
------------------------------------------------------------------------------

package body Rational_Arithmetics is

  --====================================================================
  -- Author    Christoph Grein
  -- Version   3.0
  -- Date      27 February 2006
  --====================================================================
  -- Because we've made some operators abstract, we've to use some type
  -- conversions to obtain the needed operations. They will not produce
  -- any machine code.
  -- The implementation guarantees that rational numbers are valid or
  -- else Constraint_Error will be raised.
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    1.0  30.06.2002
  --  C.G.    1.1  05.07.2002 added mixed relational operators,
  --                          Image and Value
  --  C.G.    2.0  15.11.2005 Ada 2005
  --  C.G.    2.1  27.11.2005 Denominator returns Positive_Whole
  --  C.G.    2.2  01.02.2006 Bug fix in Value
  --  C.G.    3.0  27.02.2006 pragma Pure: Put Image and Value in child
  --====================================================================

  function GCD (Left, Right: Whole) return Integer is
    L: Whole := Left;
    R: Whole := Right;
  begin
    while R /= 0 loop
      L := L mod R;
      if L = 0 then
        return Integer (R);
      end if;
      R := R mod L;
    end loop;
    return Integer (L);
  end GCD;

  function Reduce (X: Rational) return Rational is
  begin
    if X.Denominator = 0 then
      raise Constraint_Error;
    elsif X.Numerator = 0 then
      return (0, 1);
    end if;
    declare
      G: constant Integer := GCD (X.Numerator, X.Denominator);
    begin
      return (Whole (Integer (X.Numerator  ) / G),
              Whole (Integer (X.Denominator) / G));
    end ;
  end Reduce;

  function Numerator (R: Rational) return Whole is
  begin
    return R.Numerator;
  end Numerator;

  function Denominator (R: Rational) return Positive_Whole is
  begin
    return R.Denominator;
  end Denominator;

  function "+" (Right: Whole) return Rational is
  begin
    return (Right, 1);
  end "+";

  function "-" (Right: Whole) return Rational is
  begin
    return (Whole (-Integer (Right)), 1);
  end "-";

  function "/" (Left, Right: Whole) return  Rational is
    L:          Integer :=     Integer (Left);
    R: constant Integer := abs Integer (Right);
  begin
    if Right < 0 then
      L := -L;
    end if;
    return Reduce ((Whole (L), Whole (R)));
  end "/";

  function "abs" (Right: Rational) return Rational is
  begin
    return (abs Right.Numerator, Right.Denominator);
  end "abs";

  function "+" (Right: Rational) return Rational is
  begin
    return Right;
  end "+";

  function "-" (Right: Rational) return Rational is
  begin
    return (Whole (-Integer (Right.Numerator)), Right.Denominator);
  end "-";

  function "+" (Left, Right: Rational) return Rational is
    R: constant Rational :=
         (Left.Numerator   * Right.Denominator + Left.Denominator * Right.Numerator,
          Left.Denominator * Right.Denominator);
  begin
    return Reduce (R);
  end "+";

  function "-" (Left, Right: Rational) return Rational is
    R: constant Rational :=
         (Left.Numerator   * Right.Denominator - Left.Denominator * Right.Numerator,
          Left.Denominator * Right.Denominator);
  begin
    return Reduce (R);
  end "-";

  function "*" (Left, Right: Rational) return Rational is
    R: constant Rational := (Left.Numerator   * Right.Numerator  ,
                             Left.Denominator * Right.Denominator);
  begin
    return Reduce (R);
  end "*";

  function "/" (Left, Right: Rational) return Rational is
    L:          Rational :=     Left ;
    R: constant Rational := abs Right;
  begin
    if Right.Numerator < 0 then
      L.Numerator := Whole (-Integer (L.Numerator));
    end if;
    return Reduce ((L.Numerator   * R.Denominator,
                    L.Denominator * R.Numerator  ));
  end "/";

  function "=" (Left: Whole; Right: Rational) return Boolean is
  begin
    return Left = Right.Numerator and Right.Denominator = 1;
  end "=";

  function "=" (Left: Rational; Right: Whole) return Boolean is
  begin
    return Right = Left;
  end "=";

  function "<" (Left, Right: Rational) return Boolean is
  begin
    return Left.Numerator * Right.Denominator < Left.Denominator * Right.Numerator;
  end "<";

  function "<=" (Left, Right: Rational) return Boolean is
  begin
    return Left.Numerator * Right.Denominator <= Left.Denominator * Right.Numerator;
  end "<=";

  function ">=" (Left, Right: Rational) return Boolean is
  begin
    return Left.Numerator * Right.Denominator >= Left.Denominator * Right.Numerator;
  end ">=";

  function ">" (Left, Right: Rational) return Boolean is
  begin
    return Left.Numerator * Right.Denominator > Left.Denominator * Right.Numerator;
  end ">";

  function "<" (Left: Whole; Right: Rational) return Boolean is
  begin
    return +Left < Right;
  end "<";

  function "<" (Left: Rational; Right: Whole) return Boolean is
  begin
    return Left < +Right;
  end "<";

  function "<=" (Left: Whole; Right: Rational) return Boolean is
  begin
    return +Left <= Right;
  end "<=";

  function "<=" (Left: Rational; Right: Whole) return Boolean is
  begin
    return Left <= +Right;
  end "<=";

  function ">=" (Left: Whole; Right: Rational) return Boolean is
  begin
    return +Left >= Right;
  end ">=";

  function ">=" (Left: Rational; Right: Whole) return Boolean is
  begin
    return Left >= +Right;
  end ">=";

  function ">" (Left: Whole; Right: Rational) return Boolean is
  begin
    return +Left > Right;
  end ">";

  function ">" (Left: Rational; Right: Whole) return Boolean is
  begin
    return Left > +Right;
  end ">";

  function Max (Left, Right: Rational) return Rational is
  begin
    if Left >= Right then
      return Left;
    else
      return Right;
    end if;
  end Max;

  function Min (Left, Right: Rational) return Rational is
  begin
    if Left <= Right then
      return Left;
    else
      return Right;
    end if;
  end Min;

  function Max (Left: Whole; Right: Rational) return Rational is
  begin
    return Max (+Left, Right);
  end Max;

  function Max (Left: Rational; Right: Whole) return Rational is
  begin
    return Max (Left, +Right);
  end Max;

  function Min (Left: Whole; Right: Rational) return Rational is
  begin
    return Min (+Left, Right);
  end Min;

  function Min (Left: Rational; Right: Whole) return Rational is
  begin
    return Min (Left, +Right);
  end Min;

  function "+" (Left: Whole; Right: Rational) return Rational is
    R: constant Rational := (Left * Right.Denominator + Right.Numerator,
                             Right.Denominator);
  begin
    return Reduce (R);
  end "+";

  function "+" (Left: Rational; Right: Whole) return Rational is
  begin
    return Right + Left;
  end "+";

  function "-" (Left: Whole; Right: Rational) return Rational is
    R: constant Rational := (Left*Right.Denominator - Right.Numerator,
                             Right.Denominator);
  begin
    return Reduce (R);
  end "-";

  function "-" (Left: Rational; Right: Whole) return Rational is
    R: constant Rational := (Left.Numerator - Right * Left.Denominator,
                             Left.Denominator);
  begin
    return Reduce (R);
  end "-";

  function "*" (Left: Whole; Right: Rational) return Rational is
    R: constant Rational := (Left * Right.Numerator,
                             Right.Denominator);
  begin
    return Reduce (R);
  end "*";

  function "*" (Left: Rational; Right: Whole) return Rational is
  begin
    return Right * Left;
  end "*";

  function "/" (Left: Whole; Right: Rational) return Rational is
    R: constant Rational := (Left * Right.Denominator,
                             Right.Numerator);
  begin
    return Reduce (R);
  end "/";

  function "/" (Left: Rational; Right: Whole) return Rational is
    R: constant Rational := (Left.Numerator,
                             Left.Denominator * Right);
  begin
    return Reduce (R);
  end "/";

end Rational_Arithmetics;
