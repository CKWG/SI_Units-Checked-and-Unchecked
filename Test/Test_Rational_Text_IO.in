  
 8  -4   
     -21/7 3/15   
 -      exception -
 /      exception /
 - 8    exception - 8
 7 /21  exception  /21
 -21/ 7 exception -21/ 7
