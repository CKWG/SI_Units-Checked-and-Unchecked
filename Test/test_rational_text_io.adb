------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2006, 2020 Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 2
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Author's homepage and email address:
--   http://home.T-Online.de/home/Christ-Usch.Grein/
--   Christ-Usch.Grein@T-Online.de
------------------------------------------------------------------------------

with Ada.Exceptions;
use  Ada.Exceptions;

with Ada.Text_IO;
use  Ada.Text_IO;

with Test_Support;
use  Test_Support;

with Rational_Arithmetics.Text_IO;
use  Rational_Arithmetics, Rational_Arithmetics.Text_IO;

procedure Test_Rational_Text_IO is

  --====================================================================
  -- Author    Christoph Grein
  -- Version   2.0
  -- Date      24 July 2020
  --====================================================================
  -- Test the text input output facility.
  -- The program reads the file Test_Rational_Text_IO.in and writes the
  -- result to standard output. Compare the output with that in file
  -- Test_Rational_Text_IO.out.expected - it must be identical.
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    1.0  02.02.2006
  --  C.G.    2.0  24.07.2020 Compare expected and actual output
  --====================================================================

  Rat_in, Rat_out: File_Type;

  R: Rational;

  Line: String (1 .. 30);
  Last: Natural;

begin

  Test_Header (Title => "Test Rational Text_IO",
               Description => "Test reading from and writing to a file.");

  Test_Step (Title => "Produce Ouput",
             Description => "Read input file and write output file.");

  Open   (Rat_in , Name => "Test_Rational_Text_IO.in" , Mode => In_File);
  Create (Rat_out, Name => "Test_Rational_Text_IO.out", Mode => Out_File);

  while not End_Of_File (Rat_in) loop
    begin
      Get (Rat_in , R);
      Put (Rat_out, R);  New_Line (Rat_out);
    exception
      when E:others =>
        Get_Line (Rat_in , Line, Last);
        Put_Line (Rat_out, Exception_Name (E) & " => """ & Line (1 .. Last) & '"');
    end;
  end loop;

  Test_Step (Title => "Compare",
             Description => "Compare the actual with the expected output.");

  Close (Rat_in);
  Reset (Rat_out, In_File);
  Open  (Rat_in , In_File, "Test_Rational_Text_IO.out.expected");

  declare
    Line_i, Line_o: String (1 .. 55);
    Last_i, Last_o: Natural;
  begin
    loop
      Get_Line (Rat_in , Line_i, Last_i);
      Get_Line (Rat_out, Line_o, Last_o);
      Assert (Condition => Line_i (1 .. Last_i) = Line_o (1 .. Last_o),
              Message   => Line_o (1 .. Last_o),
              Only_Report_Error => False);
    end loop;
  exception
    When End_Error => null;
  end;

  Close  (Rat_in);
  Delete (Rat_out);

  Test_Result;

end Test_Rational_Text_IO;
