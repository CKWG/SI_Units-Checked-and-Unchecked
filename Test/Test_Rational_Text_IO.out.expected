 8
-4
-3
 1/5
ADA.IO_EXCEPTIONS.DATA_ERROR => "      exception -"
ADA.IO_EXCEPTIONS.DATA_ERROR => "/      exception /"
ADA.IO_EXCEPTIONS.DATA_ERROR => " 8    exception - 8"
 7
ADA.IO_EXCEPTIONS.DATA_ERROR => "/21  exception  /21"
ADA.IO_EXCEPTIONS.DATA_ERROR => " 7 exception -21/ 7"
